File: plugins/spigot/TitleManager/src/main/java/io/puharesource/mc/titlemanager/backend/reflections/ReflectionClass.java
package io.puharesource.mc.titlemanager.backend.reflections;

import lombok.SneakyThrows;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

public final class ReflectionClass {
    private final Class<?> handle;
    private final String path;

    public ReflectionClass(String path) throws ClassNotFoundException {
        handle = Class.forName(path);
        this.path = path;
    }

    public Method getMethod(String methodName, Class<?>... params) throws NoSuchMethodException {
        main:
        for (Method method : handle.getDeclaredMethods()) {
            if (!method.getName().equals(methodName) || method.getParameterTypes().length != params.length) continue;
            for (int i = 0; method.getParameterTypes().length > i; i++)
                if (!method.getParameterTypes()[i].equals(params[i])) continue main;
            return method;
        }

        throw new NoSuchMethodException("Couldn't find method \"" + methodName + "\" for " + handle.getName() + ".");
    }

    public Constructor getConstructor(Class<?>... params) throws NoSuchMethodException {
        main:
        for (Constructor constructor : handle.getDeclaredConstructors()) {
            if (constructor.getParameterTypes().length != params.length) continue;
            for (int i = 0; constructor.getParameterTypes().length > i; i++)
                if (constructor.getParameterTypes()[i] != params[i]) continue main;
            return constructor;
        }

        throw new NoSuchMethodException("Couldn't find constructor for " + handle.getName() + ".");
    }

    @SneakyThrows
    public Object createInstance(Object... objects) {
        Class<?>[] classes = new Class<?>[objects.length];

        for (int i = 0; objects.length > i; i++) {
            classes[i] = objects[i].getClass();
        }

        return getConstructor(classes).newInstance(objects);
    }

    @SneakyThrows(ClassNotFoundException.class)
    public Class<?> getInnerClass(final String className) {
        return Class.forName(path + "$" + className);
    }

    @SneakyThrows(ClassNotFoundException.class)
    public ReflectionClass getInnerReflectionClass(final String className) {
        return new ReflectionClass(path + "$" + className);
    }

    @SneakyThrows(NoSuchFieldException.class)
    public Field getField(final String name) {
        return handle.getField(name);
    }

    public final Class<?> getHandle() {
        return handle;
    }
}

File: src/main/kotlin/io/puharesource/mc/titlemanager/internal/reflections/ReflectionClass.kt
package io.puharesource.mc.titlemanager.internal.reflections

import java.lang.reflect.Constructor
import java.lang.reflect.Field

class ReflectionClass(private val path: String) {
    val handle: Class<*> = Class.forName(path)

    fun getMethod(methodName: String, vararg params: Class<*>) = handle.getDeclaredMethod(methodName, *params)
        ?: throw NoSuchMethodException("Couldn't find method for ${handle.name}.")

    fun getConstructor(vararg params: Class<*>) = handle.getDeclaredConstructor(*params)
        ?: throw NoSuchMethodException("Couldn't find constructor for ${handle.name}")

    fun getConstructorByParameterLength(parameterCount: Int): Constructor<*> {
        val constructor = handle.declaredConstructors.firstOrNull { it.parameterCount == parameterCount } ?: throw NoSuchMethodException("Couldn't find constructor for ${handle.name}")

        constructor.isAccessible = true

        return constructor
    }

    fun createInstance(vararg objects: Any): Any {
        val classes: Array<Class<*>> = objects.map { it.javaClass }.toTypedArray()

        return getConstructor(*classes).newInstance(*objects)
    }

    fun getField(fieldName: String): Field = handle.getDeclaredField(fieldName)

    fun getInnerClass(className: String): Class<*> = Class.forName("$path$$className")

    fun getInnerReflectionClass(className: String) = ReflectionClass("$path$$className")
}
